internal/lang/eval.go

```diff
@@ -455,13 +459,31 @@ func (s *Scope) evalContext(refs []*addrs.Reference, selfAddr addrs.Referenceabl
 		vals["output"] = cty.ObjectVal(outputValues)
 	}
 
+	if len(runBlockValue) > 0 {
+		vals["run"] = cty.ObjectVal(buildResourceObjects(runBlockValue))
+	}
+
 	if self != cty.NilVal {
 		vals["self"] = self
 	}
 
 	return ctx, diags
 }
 
+func getRunBlockValue(ref *addrs.Reference, value cty.Value, runBlockValue map[string]map[string]cty.Value) {
```

Comment:
`getRunBlockValue()` is used to store the output value with respect to the run block nested structure. for eg: if provider is trying to access the value of `run.test_module.test_value` then we need to store the value following the same nested structure in the [hcl.EvalContext](https://github.com/opentofu/opentofu/blob/17bc1acfc2678db707e005d5ecf32e5cc881ddc6/internal/lang/eval.go#L247) so that in [hcldec.Decode()](https://github.com/opentofu/opentofu/blob/17bc1acfc2678db707e005d5ecf32e5cc881ddc6/internal/lang/eval.go#L73) can evaluate it’s value based on the provided [hcl.body](https://github.com/opentofu/opentofu/blob/17bc1acfc2678db707e005d5ecf32e5cc881ddc6/internal/lang/eval.go#L52).

Reply:
Thanks for the explanation! Also, there are some confusing parts:

* Why do we check only the first and second elements of `ref.Remaining` slice? Do we need to add a length check at least?

* In the case of type cast failure we will just write the value to empty keys: `runBlockValue[block][outputVariable] = value`, where `block` and/or `outputVariable` are empty . Is it expected?

* The function is writing its return value into the argument passed. Can we refactor it a bit to follow more common practices (read-only arguments with return values)? The signature could be similar to the following: `func getRunBlockKeys(ref *addrs.Reference) (block, outputVar string)`

Reply:
We check first two elements because the we are expecting the output value referred from run block is not going to be nested further. For eg: run.setup.sample_test_value is being used in this [test file.](https://github.com/opentofu/opentofu/blob/17bc1acfc2678db707e005d5ecf32e5cc881ddc6/internal/command/testdata/test/pass_with_run_block_output/tests/main.tftest.hcl#L2)

I have added the length before calling getRunBlockValue() method [here](https://github.com/opentofu/opentofu/blob/17bc1acfc2678db707e005d5ecf32e5cc881ddc6/internal/lang/eval.go#L418).
I will update this if block to check for exact two elements as per the use case.